lib/racecar/consumer.rb

```diff
@@ -34,20 +36,46 @@ def subscribes_to(*topics, start_from_beginning: true, max_bytes_per_partition:
 
     def configure(producer:, consumer:, instrumenter: NullInstrumenter)
       @producer = producer
+      @delivery_handles = []
+
       @consumer = consumer
       @instrumenter = instrumenter
     end
 
     def teardown; end
 
-    # Delivers messages that got produced.
+    # Blocks until all messages produced so far have been successfully published. If
+    # message delivery finally fails, a Racecar::MessageDeliveryError is raised. The
+    # delivery failed for the reason in the exception. The error can be broker side
+    # (e.g. downtime, configuration issue) or specific to the message being sent. The
+    # caller must handle the latter cases or run into head of line blocking.
     def deliver!
       @delivery_handles ||= []
       if @delivery_handles.any?
         instrumentation_payload = { delivered_message_count: @delivery_handles.size }
 
         @instrumenter.instrument('deliver_messages', instrumentation_payload) do
-          @delivery_handles.each(&:wait)
+          @delivery_handles.each do |handle|
+            # rdkafka-ruby checks every wait_timeout seconds if the message was
+            # successfully delivered, up to max_wait_timeout seconds before raising
+            # Rdkafka::AbstractHandle::WaitTimeoutError. librdkafka will (re)try to
+            # deliver all messages in the background, until "config.message_timeout"
+            # (message.timeout.ms) is exceeded. Phrased differently, rdkakfa-ruby's
+            # WaitTimeoutError is just informative.
+            # The raising can be avoided if max_wait_timeout below is greater than
+            # config.message_timeout, but config is not available here (without
+            # changing the interface).
+            handle.wait(max_wait_timeout: 60, wait_timeout: 0.1)
+          rescue Rdkafka::AbstractHandle::WaitTimeoutError => e
+            delivery_report = handle.create_result
+            # ideally we could use the logger passed to the Runner, but it is not
+            # available here. The runner sets it for Rdkafka, though, so we can use
+            # that instead.
+            Rdkafka::Config.logger.debug "Still trying to deliver message to partition #{delivery_report.partition} at offset #{delivery_report.offset}... (will try up to Racecar.config.message_timeout)"
```

Comment:
> We can of course reduce the likelihood of that happening by changing the signature of `configure` to `configure(producer:, consumer:, instrumenter: NullInstrumenter, logger: Racecar.config.logger)` and adapting Racecar internal calls to it. Then someone needs to call `configure` on the consumer for this to break.

I think that makes sense :+1:

Reply:
> > Also... the offset being reported is surely not from the produced message, right?
> 
> It's good that you bring that up, it's very likely always some -1 default value. It _will_ be from the produced message, but most likely only when it actually gets delivered. All these fields get initialized to -1 from rdkafka:
> https://github.com/appsignal/rdkafka-ruby/blob/c036e1ee1173b7547ea194647550592bdfac3469/lib/rdkafka/producer.rb#L122-L126
> It then hands a pointer to the handle to librdkafka which will eventually fill in all the details. Potentially the partition can be filled even if the delivery fails, since partitioning is done client side (at least potentially, librdkafka has code for it and the [on wire protocol](http://kafka.apache.org/0101/protocol.html#protocol_messages) seems to allow it).
> 
> So long story short, the offset is very likely always -1, and I assume the partition will also be. I'll remove all the extra output, since it's very unlikely it will be helpful.

The partition is _always_ set by the client, but it may not be present on that object – although I would guess that it would be. Perhaps keep partition and we can change it later if it's empty in practice?